pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go

```diff
@@ -0,0 +1,797 @@
+					ic := &scpb.IndexColumn{
+						TableID:       tbl.TableID,
+						IndexID:       newPrimaryIndex.IndexID,
+						ColumnID:      colID,
+						OrdinalInKind: uint32(i),
+						Kind:          scpb.IndexColumn_KEY,
+						Direction:     col.Direction.ToIndexColumnDirection(),
+					}
+					b.Add(ic)
+					newColumns = append(newColumns, ic)
+					allKeyColumnIDs[colID] = true
+				}
+			}
+
+			// What's left are STORED columns, excluding virtual columns and system columns
+			i := 0
+			for _, colID := range allColumns {
+				if _, isKeyCol := allKeyColumnIDs[colID]; isKeyCol ||
+					mustRetrieveColumnTypeElem(b, tbl.TableID, colID).IsVirtual ||
+					colinfo.IsColIDSystemColumn(colID) {
+					continue
+				}
+				ic := &scpb.IndexColumn{
+					TableID:       tbl.TableID,
+					IndexID:       newPrimaryIndex.IndexID,
+					ColumnID:      colID,
+					OrdinalInKind: uint32(i),
+					Kind:          scpb.IndexColumn_STORED,
+				}
+				b.Add(ic)
+				newColumns = append(newColumns, ic)
+				i++
+			}
+
+			return newColumns
+		})
+		newPrimaryIndexElem.Sharding = makeShardedDescriptor(b, t)
+	}
+
+	// Construct and add elements for a unique secondary index created on
+	// the old primary key columns.
+	// This is a CRDB unique feature that exists in the legacy schema changer.
+	maybeAddUniqueIndexForOldPrimaryKey(b, tn, tbl, t, oldPrimaryIndexElem, newPrimaryIndexElem)
+
+	// Send a notice to users about how this job is asynchronous.
+	b.EvalCtx().ClientNoticeSender.BufferClientNotice(b,
+		pgnotice.Newf(
+			"primary key changes are finalized asynchronously; "+
+				"further schema changes on this table may be restricted until the job completes"),
+	)
+}
+
+// checkForEarlyExit asserts several precondition for a
+// `ALTER PRIMARY KEY`, including
+//   1. no expression columns allowed;
+//   2. no columns that are in `DROPPED` state;
+//   3. no inaccessible columns;
+//   4. no nullable columns;
+//   5. no virtual columns (starting from v22.1);
+//   6. add more here
+// Panic if any precondition is found unmet.
+func checkForEarlyExit(b BuildCtx, tbl *scpb.Table, t *tree.AlterTableAlterPrimaryKey) {
+	if err := paramparse.ValidateUniqueConstraintParams(
+		t.StorageParams,
+		paramparse.UniqueConstraintParamContext{
+			IsPrimaryKey: true,
+			IsSharded:    t.Sharded != nil,
+		},
+	); err != nil {
+		panic(err)
+	}
+
+	for _, col := range t.Columns {
+		if col.Column == "" && col.Expr != nil {
+			panic(errors.WithHint(
+				pgerror.Newf(
+					pgcode.InvalidColumnDefinition,
+					"expressions such as %q are not allowed in primary index definition",
+					col.Expr.String(),
+				),
+				"use columns instead",
+			))
+		}
+
+		colElems := b.ResolveColumn(tbl.TableID, col.Column, ResolveParams{
+			IsExistenceOptional: false,
+			RequiredPrivilege:   privilege.CREATE,
+		})
+
+		colCurrentStatus, _, colElem := scpb.FindColumn(colElems)
+		if colElem == nil {
+			panic(fmt.Sprintf("programming error: resolving column %v does not give a "+
```

Comment:
I'd rather you used an `errors.AssertionFailedf` error because it'll capture the stack trace. Maybe we need a helper `panicf` which combines `panic` and `errors.AssertionFailedWithDepthf`

Reply:
The other advantage of the error constructors is they will properly handle redaction management for arguemnts. If you don't construct an error and instead format a string, the higher levels will redact the entire message.